Split out CommandLine enum into a separate static library, allow removing it from stdlib

kubamracek opened the pull request:
This is for the 'freestanding' build to stop assuming the platform has argc/argv.

- Introduce a new sub-library, libswiftCommandLineSupport.a
- Move stubs/CommandLine.cpp into this library
- Conditionally embed it into libswiftCore
- Conditionally embed it into libswiftPrivateLibcExtras if not in libswiftCore to support testing
- Add SWIFT_STDLIB_HAS_COMMANDLINE CMake (and build-script) flag

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test

kubamracek commented:
@swift-ci please test

kubamracek commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

compnerd commented on lib/Immediate/Immediate.cpp:
```diff
@@ -238,7 +238,7 @@ int swift::RunImmediately(CompilerInstance &CI,
   // racing with the static initializers in user code.
   // Setup interpreted process arguments.
-  using ArgOverride = void (*)(const char **, int);
+  using ArgOverride = void (* SWIFT_CC(swift))(const char **, int);
```
Nice find - I suspect that the WASM folks would  appreciate this too

compnerd reviewed (commented):
I don't see anything worrisome except for the potential ABI issues.  I'd like to get @mikeash and @jckarter to have a look over this as well.

kubamracek committed:
Split out CommandLine enum into a separate static library, allow removing it from stdlib

This is for the 'freestanding' build to stop assuming the platform has argc/argv.

- Introduce a new sub-library, libswiftCommandLineSupport.a
- Move stubs/CommandLine.cpp into this library
- Conditionally embed it into libswiftCore
- Conditionally embed it into libswiftPrivateLibcExtras if not in libswiftCore to support testing
- Add SWIFT_STDLIB_HAS_COMMANDLINE CMake (and build-script) flag

mikeash commented:
Oh right. I somehow got it in my head that the CommandLine enum was in the static library. You’re right, no module names at play here, and it’s all good.

kubamracek commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset
